README
SMail Archiver 0.4 is a python script that can be used to backup an email account
It uses the pycrypto library and requires python 2.7 or above.

Backup:
    The emails are fetched using IMAP. UID of emails are used to identify emails.
    Each mail is saved in mbox format using the format {user}/{uid}.{ext}.
    {ext} represent the options enable and successive operations.
    {ext} of backup email always starts with '.mbox'
    If the option {compress} is used, the email will be compressed using GZip
    and the '.gz' extension is added.
    If the option {encrypt} is used, the email will be encrypted using AES
    and the '.enc' extension is added.
    A compressed and encrypted email will then have the extension .mbox.gz.enc.
    Compression is always realised before encryption for better results.
    Running the command several time will not overwrite the files, existing mbox
    files are ignored.

Restore:
    When restoring a backup, the script read the content of the specified backup
    folder and, for each file, applies the adequate operations according the
    extension in reverse order.
    '.enc' extension will apply a decryption of the file
    '.gz' extension will apply a decompression of the file
    '.mbox' extension is mandatory to indicate thr file is an email.

    The result of all valid restorations is stored in a mbox file using the
    format {foldername}.mbox, overwritting potential previous restoration.
    
    Files contained in a folder are processsed in an arbitrary order (not
    alphabetical).
    
Security:
    The script uses AES256 in CBC mode for encryption and HMAC-SHA256 for 
    signature. The size of the AES key can be changed with the variable 
    AES_KEY_SIZE (16, 24, 32). Use a different algorithms is not proposed as
    requires more change in the code.
    
    Two modes of security are possible : simple or password-protected.
    In simple mode (the variable {promp} is not set), no password from the user
    is required, the two (AES and HMAC) keys are generated using a PRNG and are
    the secret. The generated file SHOULD be protected in a secure way.
    In password protected mode (the variable {promp} is true), a password is
    asked and the two keys are generated using this password and random salt
    using PBKDF2. The generated file contains the salt and should no longer be
    protected. However, the strength of the encryption lies in the strength of 
    the password and it could be an additionnal protection to protect the salt
    file in case of weak password.

    The result of encryption is saved in BASE64 format.

Usage:
    usage: smailarchiver [-h] [-u USER] [-i IMAP] [-p PASSWD] [-f FOLDER] [-e]
                         [-k KEYS] [-P] [-z] [-c CONFIG] [-r RESTORE] [--verbose]
                         [--version]
    
    optional arguments:
      -h, --help            show this help message and exit
      -u USER, --user USER  email username
      -i IMAP, --imap IMAP  IMAP email server
      -p PASSWD, --passwd PASSWD
                            email password
      -f FOLDER, --folder FOLDER
                            inbox folder name
      -e, --encrypt         enable encryption
      -k KEYS, --keys KEYS  key/salt file (ignored if encryption disabled)
      -P, --promp           promp for password of keys (ignored if encryption
                            disabled)
      -z, --compress        Compress the mail data
      -c CONFIG, --config CONFIG
                            JSON file containing a list of configuration
      -r RESTORE, --restore RESTORE
                            restore the mails contained in a folder. Extension is
                            used to determine the storage type. Restored in a
                            single .mbox file.
      --verbose             verbose mode
      --version             show program's version number and exit
    
    The optional -c CONFIG uses a JSON file containing the configuration (useful
    for cronjob for instance). It can contain the configuration for several 
    mail accounts. An example of the format used is present in 
    'config.json.example'.

    The option -r RESTORE requires the name of the folder containing the backup
    files. The restored emails will be stored in the {RESTORE}.mbox file.

IMAP folder:
    The processed folder is specified in the variable {folder}. By default
    "INBOX" is used which should be available by many email providers.
    
    In the case of Gmail, all emails can be saved easily using the name 
    '"[Gmail]/All Mail"'. Note that this name can be change depending of the
    language of your email account or not available in IMAP. Read the following
    page for more information :
    http://www.sebsauvage.net/wiki/doku.php?id=gmail_archiver
    
    When using a folder name with spaces (like in the Gmail example above), you
    may need to double-quote the name ('"folder name"' or "'folder name'") in
    Python 3.x

    To know which IMAP folders are available, open a python promp and use the 
    following code :

    >>> import imaplib
    >>> m = imaplib.IMAP4_SSL("imap.bar.com")
    >>> m.login("foo@bar.com","mypasword")
    ('OK', [b'Logged in'])
    >>> m.list()
    ('OK', [b'(\\HasNoChildren) "." "Junk"', b'(\\HasNoChildren) "." "Trash"', b
    '(\\HasNoChildren) "." "Drafts"', b'(\\HasNoChildren) "." "Sent"', b'(\\HasN
    oChildren) "." "INBOX"'])

    (or `m.list("foldername")` for the subfolders)
    
    You can specify only one IMAP folder per configuration. However, you can
    configure several sites for the same email account in the the config.json
    file. Note that the destination folder is specified using the {user} 
    variable and emails from the same account will be merged. Emails are
    identified with UID which are unique per message. Emails with the same UID
    in the destination folder ({UID}.mbox.{ext} depending of the options) are
    skipped during the save process. At the restauration process, all emails in
    the specified folder will be merged in a {user}.mbox file.

Update:
    When you update SMailArchiver, look out for incompatibilities.
    The file 'migration.py' is a script of migration of stored files to new
    formats. Look at the content of this file for more explanations how to
    migrate your stored emails.


CHANGELOG
2012-01-07 0.4
    * --decrypt option becomes --restore
    * extension file makes more sense (.mbox.gz.enc means mbox -> compression -> encryption)
    * encryption is now optional
    * config file takes arguments 'encrypt' and 'compress'
    * license use MIT license

2012-12-19 0.3
    * Initial functional release
    * Encryption and decrytion
    * Compression and decompression
    * Config file



migration.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""SMailArchiver Migration tool

Between versions of SMailArchiver, the file format used may change for diverse
reasons. To ensure you will still be able to encrypt/recieved your emails after
upgrading the script, this tool was developped.

requirements: 
  python 2.7 or above

usage:
  Indicate the version you come from and the folder you want to migrate
  $ python migrate.py -v 0.3 -f foo@bar.com/

If you skipped several versions, run several times the script for each version
eg: if a format change was integrated between version 0.3 and 0.4 and between 
0.5 and 06, you will need to run twice the program
  $ python migrate -v 0.3 -f boo@bar.com
  Migrating from 0.3 to 0.5...
  done.
  $ python migrate -v 0.5 -f boo@bar.com
  Migrating from 0.5 to 0.6...
  done.

Break in compatibility history:
  * 0.3 -> 0.4

Copyright (C) 2013 Martin Trigaux - MIT license
"""

import argparse
import json
import os

BREAK_HISTORY = [0.3]
LATEST_VERSION = 0.4


def migrate_folder(from_ver, foldername):
    
    if not os.path.isdir(foldername):
        raise OSError("Folder {} does not exists".format(foldername))

    if from_ver >= LATEST_VERSION:
        print("You have already the latest version")

    if from_ver <= 0.3:
        migrate_03(foldername)

def migrate_03(foldername):
    """Migrate from version 0.3

    In 0.3 the mails where always encrypted and used the following format:
    {UID}.mbox -> encrypted file -> {UID}.mbox.enc
    {UID}.gz.mbox -> compressed & encrypted file -> {UID}.mbox.gz.enc
    """

    print("Migrating from 0.3 to 0.4...")

    for filename in os.listdir(foldername):
        abspath = os.path.join(foldername,filename)
        if filename[-5:] == ".json":
            try:
                with open(abspath,'r') as f:
                    config_dic = json.load(f)
                assert 'list' in config_dic, "missing argument 'list'"
            except:
                # unvalid json skipping
                pass
            
            # now migration
            for config in config_dic['list']:
                config['encrypt'] = 1
                config['compress'] = 1

            with open(abspath,'w') as f:
                json.dump(config,f,sort_keys=True,indent=4)
                
            print("config file {} migrated".format(filename))

        elif filename[-5:] == ".mbox":
            if filename[-8:] == ".gz.mbox":
                destfile = "{}.mbox.gz.enc".format( filename[:-8] )
            else:
                destfile = "{}.enc".format( filename )
            
            if os.path.exists(destfile):
                raise OSError("Conflict for {0}! Destination file {1} already exists".format(filename, destfile))
            else:
                print("{0} -> {1}".format(filename,destfile))
                os.rename(abspath, os.path.join(foldername,destfile))

    print("Done.")


if __name__ == "__main__":

    parser = argparse.ArgumentParser(prog='smailmigrate')
    parser.add_argument("-v", "--version", help="the version you are coming FROM")  # eg: 0.3
    parser.add_argument("-f", "--folder", help="folder to migrate")  # eg: "foo@bar.com/"

    args = parser.parse_args()

    from_ver = float(args.version)
    migrate_folder(from_ver, args.folder)


smailarchiver.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""SMail archiver 0.4

Secured email archiver
Backups can be done using AES256 and
signed using HMAC-SHA256
Compression using gzip

requirements: 
  python 2.7 or above
  Pycrypto

Copyright (C) 2013 Martin Trigaux - MIT license
"""

import argparse
import base64
import getpass
import imaplib
import json
import os
import sys
import zlib

# pycrypto
from Crypto import Random
from Crypto.Cipher import AES
from Crypto.Hash import SHA256
from Crypto.Hash.HMAC import HMAC
from Crypto.Protocol.KDF import PBKDF2

if sys.version < '3':
    # in python 3 raw_input does not exist anymore
    input = raw_input

VERSION = "0.4"
AES_BLOCK_SIZE = AES.block_size # 16
AES_KEY_SIZE = 32 # AES-256 (16, 24, 32)
SIG_SIZE = SHA256.digest_size # 32

def generate_new_keys(key_file="keys", password=None, enc_key_size=AES_KEY_SIZE, sig_key_size=SIG_SIZE):
    """Generate the encryption and signature key

    If a password is specified, will use it to generate the key using PBKDF2 and random salt
    If no password is specified, the key will be random bytes.
    In the second case, the key file is the only secret, it should be protected"""

    rand_bytes = Random.get_random_bytes(enc_key_size + sig_key_size)
    
    # save the salt/secret
    with open(key_file,'wb') as f:
        f.write(base64.b64encode(rand_bytes))
        
    if password:
        # derive the two keys using PBKDF2 and the salt
        enc_key = PBKDF2(password, rand_bytes[:enc_key_size], dkLen=enc_key_size)
        sig_key = PBKDF2(password, rand_bytes[enc_key_size:], dkLen=sig_key_size)
        hash_string = base64.b64encode(SHA256.new(rand_bytes+password.encode()).digest())
        with open(key_file+".sha",'wb') as f:
            f.write(hash_string)

        print("Enc salt: {0}\nSig salt: {1}\nHash: {2}".format(
            base64.b64encode(enc_key).decode(),
            base64.b64encode(sig_key).decode(),
            hash_string.decode()))

    else:
        # the keys are the random bytes
        enc_key = rand_bytes[:enc_key_size]
        sig_key = rand_bytes[enc_key_size:]
        
        print("Enc key: {0}\nSig key: {1}".format(
            base64.b64encode(enc_key).decode(),
            base64.b64encode(sig_key).decode()))
    
    return (enc_key, sig_key)


def load_keys(key_file, password=None, enc_key_size=AES_KEY_SIZE, sig_key_size=SIG_SIZE):
    """ Load the keys from the specified file
    
    If a password is specified, the key file contains the salt that is used to 
        generate the key using PBKDF2 and the password
    If no password is specified, the key will be the content of the key file"""

    with open(key_file,'r') as f:
        content = f.read()
    rand_bytes = base64.b64decode(content)
    
    assert len(rand_bytes) == enc_key_size + sig_key_size, "invalid key"
    
    if password:
        # verifiy the password
        if not os.path.isfile(key_file+".sha"):
            raise Exception("No hash file found, can not guarantee the password")
        else:
            hash_string = base64.b64encode(SHA256.new(rand_bytes+password.encode()).digest())
            with open(key_file+".sha",'r') as f:
                assert f.read().encode() == hash_string, "Hash validation failed"

        # derive the two keys using PBKDF2
        enc_key = PBKDF2(password, rand_bytes[:enc_key_size], dkLen=enc_key_size)
        sig_key = PBKDF2(password, rand_bytes[enc_key_size:], dkLen=sig_key_size)
    else:
        # the keys are the random bytes
        enc_key = rand_bytes[:enc_key_size]
        sig_key = rand_bytes[enc_key_size:]

    return (enc_key, sig_key)


def encrypt(data, aes_key, hmac_key, enc_key_size=AES_KEY_SIZE, enc_block_size=AES_BLOCK_SIZE, sig_key_size=SIG_SIZE):
    """encrypt data with AES-CBC and sign it with HMAC-SHA256"""
    pad = enc_key_size - len(data) % sig_key_size
    data = data + pad * chr(pad).encode()
    iv_bytes = Random.get_random_bytes(enc_block_size)
    cypher = AES.new(aes_key, AES.MODE_CBC, iv_bytes)
    enc_data = iv_bytes + cypher.encrypt(data)
    sig = HMAC(hmac_key, enc_data, SHA256).digest()
    return enc_data + sig

def decrypt(data, aes_key, hmac_key, enc_key_size=AES_KEY_SIZE, enc_block_size=AES_BLOCK_SIZE, sig_key_size=SIG_SIZE):
    """Verify HMAC-SHA256 signature and decrypt data with AES-CBC"""
    #aes_key, hmac_key = keys
    data = base64.b64decode(data)
    sig = data[-sig_key_size:]
    data = data[:-sig_key_size]
    
    assert HMAC(hmac_key, data, SHA256).digest() == sig, "Invalid data signature"

    iv_bytes = data[:enc_block_size]
    data = data[enc_block_size:]
    cypher = AES.new(aes_key, AES.MODE_CBC, iv_bytes)
    data = cypher.decrypt(data)
    
    # eg: data = b'...\x03\x03\x03'
    if sys.version > '3':
        # data[-1] = 3, int
        return data[:-data[-1]]
    else:
        # data[-1]  = '\x03', str
        return data[:-ord(data[-1])]


def restore_folder(foldername, key_file, promp):
    """Restore the content of foldername to a single mbox file

    Each mbox file in the folder is read and, if needed, decrypted and/or 
    decompressed in the reverse order of extension
    eg: 1.mbox.gz.enc -> read -> decryption -> decompression -> write
    supported :
        .mbox
        .mbox.gz
        .mbox.gz.enc
        .mbox.enc
    The result is writen in a {foldername}.mbox file"""

    if promp:
        passwd = getpass.getpass("Enter your encryption/signature password: ")
    else:
        passwd = None

    if not os.path.isfile(key_file):
        raise OSError("Keyfile {} does not exists".format(key_file))
    enc_key, sig_key = load_keys(key_file, passwd)
    
    if not os.path.isdir(foldername):
        raise OSError("Folder {} does not exists".format(foldername))
    
    out_file = open(os.path.abspath(foldername)+'.mbox','wb') # foo@bar.com/ -> foo@bar.com.mbox
    for filename in os.listdir(foldername):
        if filename[-9:] == ".mbox.enc" or filename[-12:] == ".mbox.gz.enc" or \
            filename[-8:] == ".mbox.gz" or filename[-5:] == ".mbox":
            with open(os.path.join(foldername,filename),'r') as mbox_file:
                file_content = mbox_file.read()

        if filename[-9:] == ".mbox.enc" or filename[-12:] == ".mbox.gz.enc":
            print("Decrypting {}".format(filename))
            file_content = decrypt(file_content, enc_key, sig_key)
            
        if filename[-8:] == ".mbox.gz" or filename[-12:] == ".mbox.gz.enc":
            print("Decompressing {}".format(filename))
            file_content = zlib.decompress(file_content)

        if file_content:
            out_file.write(file_content)
            out_file.write(b"\n\n")

        file_content = ""

    out_file.close()


def load_configs(filename):
    """Load a JSON config file and check the validity

    The format is shown in config.json.example file"""

    if not os.path.isfile(filename):
        raise OSError("{} does not exists".format(filename))

    with open(filename, 'r') as f:
        config_dic = json.load(f)

    if 'verbose' in config_dic:
        config_dic['verbose'] = bool(config_dic['verbose'])
    
    assert 'list' in config_dic, "missing argument 'list', no backup configuration specified "
    
    for config in config_dic['list']:
        assert 'user' in config, "missing argument 'user', no imap username specified"
        assert 'imap' in config, "missing argument 'imap', no imap server specified"
        assert 'passwd' in config, "missing argument 'passwd', no email password specified"
        assert 'folder' in config, "missing argument 'folder', no inbox folder specified"
        
        assert 'encrypt' in config, "missing argument 'encrypt', should the data be encrypted"
        config['encrypt'] = bool(config['encrypt'])
        if config['encrypt']:
            assert 'keys' in config, "missing argument 'keys', no keyfile specified for encryption"
            assert 'promp' in config, "missing argument 'promp', does the encryption keyfile needs a password ?"
            config['promp'] = bool(config['promp'])
        
        assert 'compress' in config, "missing argument 'compress', should the data be compressed ?"
        config['compress'] = bool(config['compress'])

    return config_dic

class EmailBackup():

    def __init__(self, user, imap, passwd, encrypt=True, compress=True, verbose=False):
        self.user = user
        self.imap = imap
        self.passwd = passwd
        self.encrypt = encrypt
        self.compress = compress
        self.verbose = verbose

        self.m = imaplib.IMAP4_SSL(imap)
        self.m.login(user,passwd)

    def get_mail_list(self,mailbox='INBOX'):
        """Return the list of UID in the mailbox folder"""
        status, msg = self.m.select(mailbox, True)
        if status != 'OK':
            raise Exception(msg[0])
    
        resp, items = self.m.uid('search', None, "ALL")
        self.items = items[0].split()
        if self.verbose: print("Found {0} emails.".format(len(self.items)))

    def get_crypto_keys(self, key_file, promp=False):
        """Load the encryption and signature key

        If no key exists, generate them
        If prop is True, will use a submited password to protect the key"""
        if promp:
            pwd = getpass.getpass("Enter your encryption/signature password: ")
        else:
            pwd = None

        # get aes keys
        if os.path.isfile(key_file):
            self.enc_key, self.sig_key = load_keys(key_file, pwd)
        else:
            self.enc_key, self.sig_key = generate_new_keys(key_file, pwd)

    def get_items(self):
        """Fetch and process each email in self.items
        
        Depending of the arguments the files will be compressed or/and encrypted
        The files' extensions depends of the parameters
        .mbox is always used (initial mail format when retrieved)
        .gz is added if compression is enabled
        .enc is added if encryption is enabled
        compression is always done before encryption (better results)
        a compressed and encypted mail will be stored as file.mbox.gz.enc"""
        
        # prepare the path
        foldername = self.user
        if os.path.isfile(foldername):
            raise OSError("a file with the same name as the desired dir, '%s', already exists." % newdir)

        if not os.path.isdir(foldername):            
            os.makedirs(foldername,mode=0o777)

        # get the items
        count = len(self.items)
        for email_uid in self.items:

            filename = os.path.join(foldername,"{}.mbox".format(email_uid.decode()))
            if self.compress:
                # compressed file has a .gz extension
                filename = "{}.gz".format(filename)
            if self.encrypt:
                # encrypted file has a .enc extension
                filename = "{}.enc".format(filename)

            if os.path.isfile(filename):
                if self.verbose: print("Skipping email {0} ({1} remaining)".format(email_uid.decode(),count))
                
            else:
                if self.verbose: print("Downloading email {0} ({1} remaining)".format(email_uid.decode(),count))
                result, data = self.m.uid('fetch', email_uid, '(RFC822)')
                email_body = data[0][1]
                # We duplicate the From: line to the beginning of the email because mbox format requires it.
                from_line = "from:unknown@unknown"
                try:
                    from_line = [line for line in email_body[:16384].split(b'\n') if line.lower().startswith(b'from:')][0].strip()
                except IndexError:
                    print("  'from:' unreadable.")
            
                email_body = b"From "+from_line[5:].strip()+b"\n"+email_body
                
                if self.compress:
                    email_body = zlib.compress(email_body)

                if self.encrypt:
                    email_body = base64.b64encode(encrypt(email_body, self.enc_key, self.sig_key))

                with open(filename,'wb') as f:
                    f.write(email_body)

            count -= 1


if __name__ == "__main__":

    parser = argparse.ArgumentParser(prog='smailarchiver')
    parser.add_argument("-u", "--user", help="email username")  # "foo@bar.com"
    parser.add_argument("-i", "--imap", help="IMAP email server") # "imap.bar.com"
    parser.add_argument("-p", "--passwd", help="email password") # "password"
    parser.add_argument("-f", "--folder", help="inbox folder name", default="INBOX") # '"[Gmail]/All Mail"'

    parser.add_argument("-e", "--encrypt", help="enable encryption", action="store_true")
    parser.add_argument("-k", "--keys", help="key/salt file (ignored if encryption disabled)", default="keys")
    parser.add_argument("-P", "--promp", help="promp for password of keys (ignored if encryption disabled)", action="store_true")

    parser.add_argument("-z", "--compress", help="Compress the mail data", action="store_true")

    parser.add_argument("-c", "--config", help="JSON file containing a list of configuration")
    
    parser.add_argument("-r", "--restore", help="restore the mails contained in a folder. "\
                            "Extension is used to determine the storage type. "\
                            "Restored in a single .mbox file.")
    
    parser.add_argument("--verbose", help="verbose mode", action="store_true")
    parser.add_argument('--version', action='version', version='%(prog)s {}'.format(VERSION))

    args = parser.parse_args()

    if args.restore:
        restore_folder(args.restore, args.keys, args.promp)
    else:

        if args.config:
            configs = load_configs(str(args.config))
            for config in configs['list']:
                print("Backup for {0}@{1}".format(config['user'],config['imap']))
                eb = EmailBackup(config['user'],config['imap'],config['passwd'],
                                 config['encrypt'],config['compress'],
                                 configs['verbose'])
                if configs['verbose']: print("Get mail list")
                eb.get_mail_list(str(config['folder']))
                
                if config['encrypt']:
                    if configs['verbose']: print("Get crypto keys")
                    eb.get_crypto_keys(config['keys'],config['promp'])
                
                if configs['verbose']: print("Get items")
                eb.get_items()

        else:

            if args.user:
                user = str(args.user)
            else:
                user = input("Enter your email username: ")
            if args.imap:
                imap = str(args.imap)
            else:
                imap = input("Enter your imap server: ")
            if args.passwd:
                pwd = str(args.passwd)
            else:
                pwd = getpass.getpass("Enter your password: ")

            if args.verbose: print("Backup for {0}@{1}".format(user,imap))
            eb = EmailBackup(user, imap, pwd,
                             args.encrypt, args.compress,
                             args.verbose)
            if args.verbose: print("Get mail list")
            eb.get_mail_list(str(args.folder))
            
            if args.encrypt:
                if args.verbose: print("Get crypto keys")
                eb.get_crypto_keys(args.keys, args.promp)
            
            if args.verbose: print("Get items")
            eb.get_items()
